File: hall-api/src/main/scala/nirvana/hall/api/internal/ExceptionUtil.java
package nirvana.hall.api.internal;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.io.Writer;

public class ExceptionUtil {

    /**
     * 获取异常信息
     * @param e 异常
     */
    public static String getStackTraceInfo(Exception e) {
        String info = "";
        Writer writer = null;
        PrintWriter printWriter = null;
        try {
            writer = new StringWriter();
            printWriter = new PrintWriter(writer);
            e.printStackTrace(printWriter);
            info = writer.toString();
//            if(info.length() > 4000){
//                info = info.substring(0, 4000);
//            }
        } catch (Exception exception) {
            exception.printStackTrace();
        } finally {
            if (printWriter != null) {
                printWriter.close();
            }
            if (writer != null) {
                try {
                    writer.close();
                } catch (IOException ioException) {
                    ioException.printStackTrace();
                }
            }
        }
        return info;
    }

}

File: hall-api/src/main/scala/nirvana/hall/api/internal/ExceptionUtil.scala
package nirvana.hall.api.internal

import java.io.{PrintWriter, StringWriter}

/**
  * Created by yuchen on 2017/6/10.
  */
object ExceptionUtil {

  /**
    * 获取异常信息
    *
    * @param e 异常
    */
  def getStackTraceInfo[A<:Throwable](e: A): String = {
    var info = ""
    val writer = new StringWriter
    val printWriter = new PrintWriter(writer,true)
      try{
        e.printStackTrace(printWriter)
        info = writer.toString
        printWriter.flush
        writer.flush
      }
    catch {
      case exception: Exception => {
        exception.printStackTrace()
      }
    } finally {
      if (printWriter != null) printWriter.close()
      if (writer != null) writer.close()
    }
    info
  }
}
